dc/s2n-quic-dc/src/socket/recv/pool.rs

```diff
@@ -0,0 +1,492 @@
+                // initialize the address
+                addr.write(Addr::default());
+                // initialize the descriptor - note that it is self-referential to `addr`, `data`, and `free`
+                // SAFETY: address, payload, and memory are all initialized
+                descriptor.write(DescriptorInner::new(
+                    idx as _,
+                    max_packet_size,
+                    NonNull::new_unchecked(addr),
+                    NonNull::new_unchecked(data),
+                    self.free.clone(),
+                ));
+
+                // push the descriptor into the free list
+                let descriptor = Descriptor::new(NonNull::new_unchecked(descriptor));
+                pending.push(descriptor);
+            }
+        }
+
+        self.free.record_region(region, pending);
+    }
+}
+
+impl Drop for Pool {
+    #[inline]
+    fn drop(&mut self) {
+        let _ = self.free.close();
+    }
+}
+
+struct Region {
+    ptr: NonNull<u8>,
+    layout: Layout,
+}
+
+struct RegionLayout {
+    packet: Layout,
+    addr_offset: usize,
+    packet_offset: usize,
+    max_packet_size: u16,
+}
+
+unsafe impl Send for Region {}
+unsafe impl Sync for Region {}
+
+impl Region {
+    #[inline]
+    fn alloc(mut max_packet_size: u16, packet_count: usize) -> (Self, RegionLayout) {
+        debug_assert!(max_packet_size > 0, "packets need to be at least 1 byte");
+        debug_assert!(packet_count > 0, "there needs to be at least 1 packet");
+
+        // first create the descriptor layout
+        let descriptor = Layout::new::<DescriptorInner>();
+        // extend it with the address value
+        let (header, addr_offset) = descriptor.extend(Layout::new::<Addr>()).unwrap();
+        // finally place the packet data at the end
+        let (packet, packet_offset) = header
+            .extend(Layout::array::<u8>(max_packet_size as usize).unwrap())
+            .unwrap();
+
+        // add any extra padding we need
+        let without_padding_len = packet.size();
+        let packet = packet.pad_to_align();
+
+        // if we needed to add padding then use that for the packet buffer since it will go to waste otherwise
+        let padding_len = packet.size() - without_padding_len;
+        max_packet_size = max_packet_size.saturating_add(padding_len as u16);
+
+        let packets = {
+            // TODO use `packet.repeat(packet_count)` once stable
+            // https://doc.rust-lang.org/stable/core/alloc/struct.Layout.html#method.repeat
+            Layout::from_size_align(
+                packet.size().checked_mul(packet_count).unwrap(),
+                packet.align(),
+            )
+            .unwrap()
+        };
+
+        let ptr = unsafe {
+            // SAFETY: the layout is non-zero size
+            debug_assert_ne!(packets.size(), 0);
+            // ensure that the allocation is zeroed out so we don't have to worry about MaybeUninit
```

Comment:
Just to confirm: the allocation is type-stable, right? We never e.g. write/read both Addr and [u8] to the same location? If we did, then this initial initialization isn't enough because typed copies "deinit" padding bytes IIRC. But I don't think we ever do that here.

Reply:
Correct. We never use the same region for multiple types